### PR TITLE
Add a Binary Sensor Filter for state settling

### PR DESCRIPTION
**Pull request by cottsay**
# What does this implement/fix?

This filter complements the `DelayedOnOff` filter, but instead of passing the value after stabilization, this filter publishes a change in state immediately and then waits to publish any additional changes until the state remains stable for a given delay period.

The primary motivation for this filter is to reduce the latency incurred by the `Delayed*` filter family while still debouncing the sensor, at the cost of removing the noise filtering provided by the `Delayed*` filters.

I'm absolutely open to suggestions for a better name for this behavior.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3439

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [x] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
binary_sensor:
  - platform: gpio
    pin: GPIO7
    filters:
      - settle: 20ms
    id: physical_button
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by ssieb** [comment]
What if the state doesn't settle?  What if it's a stray pulse?

**Comment by cottsay** [comment]
Hi @ssieb, thanks for your interest.

> What if the state doesn't settle?

The behavior is similar to the `delayed_on_off` filter, where no states are published until the state settles. The difference, of course, is that this filter will publish the very first state change before waiting for the state to settle, while `delayed_on_off` will publish only after the state has already settled.

> What if it's a stray pulse?

I touched on this briefly in the PR description:

> The primary motivation for this filter is to reduce the latency incurred by the Delayed* filter family while still debouncing the sensor, at the cost of removing the noise filtering provided by the Delayed* filters.

When comparing to `delayed_on_off`, it's a trade-off. In exchange for improving the perceived responsiveness of the device, noise like stray pulses will not be filtered. In my experience, I only use `delayed_on_off` because of its capacity to debounce, and I haven't experienced any problems with stray pulses. I would certainly not assert that everyone's experience is the same as mine, but I also have no reason to suspect that my situation is unique.

**Comment by ssieb** [comment]
The delay filters are exactly for avoiding stray pulses.  So I'm trying to understand what the purpose of this PR is.  I've never heard of anyone trying to pre-debounce before.  What would this do with a single stray pulse?  It would turn on and then what?

**Comment by cottsay** [comment]
Maybe I'm not well-versed in the terminology here, please forgive me.

At a high level, `delayed_on_off` seems to do two things well in the context of my wall switches:
1. It reduces the rapid flip-flopping (bouncing) that sometimes happens as a switch is flipped.
2. It ignores short spontaneous pulses (signal noise?) on the input which may not be related to a transition.

I have not observed any issues with (2) in my situation, so I'd like to sacrifice the ability to filter those spontaneous pulses and assume that any state activity is related to an in-progress transition. Under that assumption, there isn't any reason to wait for the state to settle to publish the change.

> I've never heard of anyone trying to pre-debounce before.

It's not the common approach, I agree. A noteworthy implementation of this strategy is referred to as "eager" debouncing by the QMK keyboard firmware. Here's a reddit post with some excellent diagrams that makes some great points about the merits of each approach: https://www.reddit.com/r/MouseReview/comments/rx35bq/mouse_debounce_detailed_explanation/

---

Two additional thoughts I'd like to share:
1. I'm certainly not suggesting that this strategy is wholly "better" than `delayed_on_off` for a typical debouncing scenario, and I don't think the default recommendations should change. In my experience the combination of LED bulb power-up delay, the relay transition time, and the debouncing delay, make my Shelly devices seem sluggish, which is why I'm prioritizing latency here.
2. There's no particular reason not to use a combination of  the `delayed_*` filters and this filter and balance the benefits of both. I could see using a `delayed_on_off` filter with a very short duration to filter most spontaneous pulses followed by this filter to ensure no additional flip-flopping occurs during the remainder of the transition.

Thank you for the questions.

File: esphome/components/binary_sensor/filter.cpp [hunk 1]
```diff
@@ -111,6 +111,23 @@ LambdaFilter::LambdaFilter(std::function<optional<bool>(bool)> f) : f_(std::move
 
 optional<bool> LambdaFilter::new_value(bool value, bool is_initial) { return this->f_(value); }
 
+optional<bool> SettleFilter::new_value(bool value, bool is_initial) {
+  if (!steady_) {
```
Comment (jesserockz) [inline esphome/components/binary_sensor/filter.cpp]:
```suggestion
  if (!this->steady_) {
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/binary_sensor/filter.cpp (hunk 1 above)
Comment (cottsay) [reply esphome/components/binary_sensor/filter.cpp]: Applied as part of 297b681. Thanks!

**Review by jesserockz — APPROVED** [review]

**Pull request merged**